### PR TITLE
Plotting - Improved custom labelling

### LINKED ISSUE #21300: Plotting - Improve the custom labelling capability
Whenever two expressions are being plotted with the same plot command, a string representation of the expression is automatically generated:

```
expr1 = (cos(x) + sin(x * pi / 4)) * exp(-x**2 / 20)
expr2 = cos(x)
p = plot(expr1, expr2, legend=True)
```

For long expression, this is quite inconvenient as the legend will eventually cover the plot. At the moment, it is possible to change the label with the following:

```
p._series[0].label = "a"
p._series[1].label = "b"
p.show()
``` 

However, this is unnecessary complicated as it requires two steps/multiple commands to get a single plot. Ideally, something similar to the following would be much appreciated:

```
plot((expr1, "a"), (expr2, "b"), legend=True)
```

### PR DESCRIPTION
**Pull request by Davide-sd**
#### Brief description of what is fixed or changed

As mentioned in #21300 , the procedure to set custom labels to the expressions being plotted
is tedious. This commit add a new optional parameter to the plot interface, allowing to quickly
set custom labels to the expressions being plotted. For example, we can now write:

```
var("x, y")
plot((cos(x), "$f_{1}$"), (sin(x), (x, -2, 2), "f2"), (x / 10, (x, -6, 6)), legend=True)
```

Here, the first expression will be plotted over the default range `(x, -10, 10)` and will have a
label containing Latex code. The second expression will be plotted over the custom range
`(x, -2, 2)` having a string label. The last expression will be plotted over the custom range
`(x, -6, 6)` having the default label set to its string representation.

Back-compatibility has been important to maintain. However, a few adjustments had to be
done. For example, previously it was possible to set the label of a single expression with the
`label` keyword:

```
plot(sin(x), label="test", legend=True)
```

With this commit, that keyword has been phased out (it doesn't do anything anymore, look at the
`_remove_label_key` function), and the docstrings has been updated with all the necessary
information to correctly set custom labels.

From an implementation point of view, the old `check_arguments` function has been deprecated:
it was easier to come up with my own implementation in order to accommodate the new custom
label option, which can be found in the `_check_arguments` function. The latter has the same
signature of the former, and returns objects in the same format (with the added label string). 
Quite a few tests have been implemented to make sure the new function produces the expected
results.

All plot functions are going to sympify the provided arguments. However, when attempting to
sympify a string, it would either return an object of type Symbol, or it would raise an error
if any strange characters are present (/, {, }, $, ...). In order to solve this problem I had to
implement the `_plot_sympify` function, which is going to recursively sympify the provided
arguments, excluding the strings.

Finally, this PR together with PR #21256 should provide an adequate basic plotting
experience.

#### References to other Issues or PRs

Fixes #21300 

#### Release Notes

* plotting
  * Added capability to quickly set custom labels when plotting multiple expressions

### CODE REVIEW

File: sympy/plotting/plot.py [hunk 1]
```diff
@@ -2400,6 +2465,13 @@ def check_arguments(args, expr_len, nb_of_free_symbols):
        >>> check_arguments([x, x**2], 1, 1)
            [(x, (x, -10, 10)), (x**2, (x, -10, 10))]
     """
+
+    SymPyDeprecationWarning(
+        feature = "check_arguments",
+        useinstead = "_check_arguments",
+        issue = 21300,
+        deprecated_since_version = "1.8.1").warn()
```
Comment (oscarbenjamin) [inline sympy/plotting/plot.py:2473]: This should be 1.9

Comment (Davide-sd) [reply sympy/plotting/plot.py:2473]: Ohh thank you! I'll fix it right away.

**Comment by Davide-sd** [comment]
@oscarbenjamin may I ask what kind of test I should run locally to catch the above errors before pushing the commits?

File: sympy/plotting/tests/test_plot.py [hunk 2]
```diff
@@ -699,3 +703,454 @@ def test_issue_20113():
...
+    r = _check_arguments(args, 3, 2)
+    assert isinstance(r, (list, tuple, Tuple)) and len(r) == 1
+    assert len(r[0]) == 6
+    assert r[0][0] == x + y
+    assert r[0][1] == cos(x + y)
+    assert r[0][2] == sin(x + y)
+    assert r[0][3] == (x, -2, 2)
+    assert r[0][4] == (y, -4, 4)
+    assert r[0][3] != r[0][4]
+    assert r[0][5] == "test"
```
Comment (oscarbenjamin) [inline sympy/plotting/tests/test_plot.py:1112]: Why not just this:
```python
assert r == [[x + y, cos(x+y), sin(x+y), (x, -2, 2), (y, -4, 4), "test"]]
```

**Comment by oscarbenjamin** [comment]
> @oscarbenjamin may I ask what kind of test I should run locally to catch the above errors before pushing the commits?

Maybe `bin/doctest` e.g.:
```
$ bin/doctest sympy/plotting/
====================================================== test process starts =======================================================
executable:         /Users/enojb/current/sympy/38venv/bin/python  (3.8.5-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       gmpy 2.0.8
numpy:              1.19.4
hash randomization: on (PYTHONHASHSEED=2482794507)

sympy/plotting/plot.py[10] .F........                                                                                       [FAIL]
sympy/plotting/experimental_lambdify.py[2] ..                                                                                 [OK]
sympy/plotting/textplot.py[1] .                                                                                               [OK]
sympy/plotting/plot_implicit.py[1] .                                                                                          [OK]

__________________________________________________________________________________________________________________________________
______________________________________________ sympy.plotting.plot.check_arguments _______________________________________________
File "/Users/enojb/current/sympy/sympy/sympy/plotting/plot.py", line 2462, in sympy.plotting.plot.check_arguments
Failed example:
    check_arguments([cos(x), sin(x)], 2, 1)
Exception raised:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/doctest.py", line 1336, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest sympy.plotting.plot.check_arguments[3]>", line 1, in <module>
        check_arguments([cos(x), sin(x)], 2, 1)
      File "/Users/enojb/current/sympy/sympy/sympy/plotting/plot.py", line 2469, in check_arguments
        SymPyDeprecationWarning(
      File "/Users/enojb/current/sympy/sympy/sympy/utilities/exceptions.py", line 183, in warn
        warnings.warn(self, stacklevel=stacklevel)
    sympy.utilities.exceptions.SymPyDeprecationWarning: 

    check_arguments has been deprecated since SymPy 1.8.1. Use
    _check_arguments instead. See
    https://github.com/sympy/sympy/issues/21300 for more info.

**********************************************************************
File "/Users/enojb/current/sympy/sympy/sympy/plotting/plot.py", line 2465, in sympy.plotting.plot.check_arguments
Failed example:
    check_arguments([x, x**2], 1, 1)
Exception raised:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/doctest.py", line 1336, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest sympy.plotting.plot.check_arguments[4]>", line 1, in <module>
        check_arguments([x, x**2], 1, 1)
      File "/Users/enojb/current/sympy/sympy/sympy/plotting/plot.py", line 2469, in check_arguments
        SymPyDeprecationWarning(
      File "/Users/enojb/current/sympy/sympy/sympy/utilities/exceptions.py", line 183, in warn
        warnings.warn(self, stacklevel=stacklevel)
    sympy.utilities.exceptions.SymPyDeprecationWarning: 

    check_arguments has been deprecated since SymPy 1.8.1. Use
    _check_arguments instead. See
    https://github.com/sympy/sympy/issues/21300 for more info.


====================================== tests finished: 13 passed, 1 failed, in 4.82 seconds ======================================
DO *NOT* COMMIT!
```

**Comment by oscarbenjamin** [comment]
I'm not sure we need the whole deprecation process for `check_arguments`. Downstream code shouldn't be using undocumented functions like that.

File: sympy/plotting/tests/test_plot.py (hunk 2 above)
Comment (Davide-sd) [reply sympy/plotting/tests/test_plot.py:1112]: Eheh I added them one at a time as I was testing, then one copy-paste led to the other... I'll compact them.

**Comment by Davide-sd** [comment]
> Maybe `bin/doctest` e.g.:

That helps. Also, Sphinx test is failing, but I don't understand why and what I have to modify...

> I'm not sure we need the whole deprecation process for `check_arguments`. Downstream code shouldn't be using undocumented functions like that.

While it shouldn't, I think it's better safe than sorry...

File: sympy/plotting/tests/test_plot.py (hunk 2 above)
Comment (Davide-sd) [reply sympy/plotting/tests/test_plot.py]: Done

**Comment by oscarbenjamin** [comment]
The Sphinx build fails with these warnings:
```
/home/runner/work/sympy/sympy/sympy/plotting/plot.py:1377: MatplotlibDeprecationWarning: The 's' parameter of annotate() has been renamed 'text' since Matplotlib 3.3; support for the old name will be dropped two minor releases later.
  ax.annotate(**a)
/home/runner/work/sympy/sympy/sympy/plotting/plot.py:1502: RuntimeWarning: invalid value encountered in double_scalars
  cos_theta = dot_product / (vector_a_norm * vector_b_norm)
looking for now-outdated files... none found
/home/runner/work/sympy/sympy/sympy/plotting/plot.py:1502: RuntimeWarning: invalid value encountered in double_scalars
  cos_theta = dot_product / (vector_a_norm * vector_b_norm)
/home/runner/work/sympy/sympy/sympy/plotting/plot.py:docstring of sympy.plotting.plot.plot:32: WARNING: Inline literal start-string without end-string.
/home/runner/work/sympy/sympy/sympy/plotting/plot.py:docstring of sympy.plotting.plot.plot:32: WARNING: Inline strong start-string without end-string.
/home/runner/work/sympy/sympy/sympy/plotting/plot.py:docstring of sympy.plotting.plot.plot_parametric:19: WARNING: Bullet list ends without a blank line; unexpected unindent.
/home/runner/work/sympy/sympy/sympy/plotting/plot.py:docstring of sympy.plotting.plot.plot3d_parametric_line:18: WARNING: Inline literal start-string without end-string.
/home/runner/work/sympy/sympy/sympy/plotting/plot.py:docstring of sympy.plotting.plot.plot3d_parametric_line:18: WARNING: Inline strong start-string without end-string.
```
The first two or three of those warnings are always there (if you could fix them that would be great).

The build has failed because of the other warnings.

You can test the build yourself by doing:
```
$ cd doc
$ make html
```
https://docs.sympy.org/latest/documentation-style-guide.html

**Comment by sylee957** [comment]
Do you think that passing serial label with the main plot `(cos(x), "$f_{1}$"), (sin(x), (x, -2, 2), "f2"), ...` would be important or useful?
I'm not sure if plot label is more important than other matplotlib options like color, line style, ..., 
I first thought of updating `label=[]` as a list to implement that feature.

**Comment by Davide-sd** [comment]
> Do you think that passing serial label with the main plot `(cos(x), "$f_{1}$"), (sin(x), (x, -2, 2), "f2"), ...` would be important or useful?
> I'm not sure if plot label is more important than other matplotlib options like color, line style, ...,

The assumption at the base of all my recent PRs is that the plotting module should be able to quickly produce simple and understandable plots. In particular:
* quick: my workflow would benefit greatly if I was able to plot multiple symbolic expressions with a single command.
* simple: the plot command should be kept as simple and intuitive as possible.
* understandable: the resulting figure should provide useful information, not only to the user actually coding his problem, but also to other people (coworkers or people you would like to share your results).

Now that automatic line coloring has been solved for `MatplotlibBackend` (PR #21256 ), in my humble opinion the next most important feature to make a plot understandable is custom labeling. I'm going to provide an example.

The following expression represent the temperature distribution along a pipe having inner diameter `ri` and outer diameter `ro`. It comes out from symbolic integration.

```
from sympy import *
r, ro, ri = symbols("r, r_o, r_i")
mdot, cp, hc = symbols(r"\dot{m}, c_p, h_c")
alpha, k, L, z = symbols("alpha, k, L, z")
Tin, Pave = symbols(r"T_{in}, P_{ave}")

# here, the only variable is z 
Tf = (Tin + (Pave * L * pi * (ro**2 - ri**2) / (2 * mdot * cp)) * 
      (1 - sin(alpha * (L / 2 - z)) / sin(alpha * L / 2)) + 
      (alpha * Pave * L  / 2) * (cos(alpha * (L / 2 - z)) / sin(alpha * L / 2)) *
      ((ro**2 - ri**2) / (2 * hc * ri) - (1 / (2 * k)) * 
      ((r**2 - ri**2) / 2 + ro**2 * log(ri / r)))
     )

# numerical values
d = {
    ri: 0.2,
    ro: 0.4,
    L: 100,
    Pave: 1000,
    hc: 1,
    alpha: 0.031,
    mdot: 1,
    k: 0.2,
    cp: 15,
    Tin: 300,
}
```
I'd like to plot the temperature distributions at the inner and outer walls:

```
Twi = Tf.subs(r, ri)
Two = Tf.subs(r, ro)
plot(Twi.subs(d), Two.subs(d), (z, 0, 100), legend=True, xlabel="x [cm]", ylabel="T [K]")
```
This is the plot I'm getting:

Figure 10

This is not good at all. If I (the coder) am not able to distinguish who is what, how can I suppose anybody else will understand it? 

As you can see, custom label is definitely much more important than custom line style or custom line color. Let's suppose I could change the line style or the line color, the result wouldn't change: the plot is still unreadable.

Let me show the exact same plot with custom labeling:

```
plot((Twi.subs(d), "$T_{w, i}$"), (Two.subs(d), "$T_{w, o}$"), (z, 0, 100), 
     legend=True, xlabel="z [cm]", ylabel="T [K]")
```

Figure 11

As you can see, I'm able to understand the plot. Other users might be able to understand it as well.

> I first thought of updating `label=[]` as a list to implement that feature.

I also thought about it and gave it a try. The problem with this approach is that you have to provide a label for each expression. If there are too few or too many labels you will have to raise an error. What if you want to hide a line from the legend? You may set its label to an empty string; but then, how would you specify to keep the string representation for a specific expression? You will have to provide the string representation `str(expr)`... I think this approach is much more convoluted towards the end users.

The approach presented in this PR encapsulates the most important options to make a plot readable inside the tuples: `(expr, range, label)`. Now you might think: "I'd really like to be able to specify custom line styles and custom line colors. Do we need to extend the aforementioned tuple or create new keyword arguments `line_styles=[]`, ...?". My answer is NO: it is my opinion that the backend should be kept as simple as possible while providing the essential features to make a plot readable (if you keep adding stuffs, you end up rewriting a plot library, which begs the question "why?"). 

That's why I said in the beginning that this PR together with PR #21256 should provide an adequate basic plotting experience. If the user wants to create fancy plots, we should provide a reliable way to get the matplotlib axes, but that's a matter for another PR.

**Comment by sylee957** [comment]
But should that make the code for parsing more complicated?
If we have label as the last optional tuple argument, we have to deal with more variable length of tuple
I saw that lots of new code were written which may be related to this change.

**Comment by Davide-sd** [comment]
> But should that make the code for parsing more complicated?
> If we have label as the last optional tuple argument, we have to deal with more variable length of tuple
> I saw that lots of new code were written which may be related to this change.

You are right, most of the changes in this code deals with the last optional label in the tuple.

I spent a couple of hours trying to understand what the function `check_arguments` was doing, then I spent a few more hours trying to modify it without breaking back-compatibility, and finally after realizing that approach was not feasible I moved on to create my own functions. The fact that the parsing interface (which dates back 9 years ago, according to the history) was not tested, made all the development much more difficult, that's why I needed to add those tests.

At the end of the day:
* let's balance the code added to deal with the last optional label in the tuple: adding the functions `_is_range, _create_ranges, _check_arguments` and subtracting the old `check_arguments` function (which will eventually be removed because of the deprecation), this PR adds about 60 lines of code, most of which are comments that will eventually help a future coder with his improvements... Of course, than we need to add the code to deal with sympification, but that's a minor amount.
* the added tests assure back-compatibility. Computationally, they are a lot cheaper then all other tests because they do not create any plot objects or data series: they are just a bunch of comparisons between input and output arguments.

Overall, in my opinion this PR is a great bargain in order to have an intuitive and satisfactory plotting interface.

**Comment by jondo** [comment]
Sorry to being late here. Yesterday, I have suggested a different API for this functionality in #22609, which also feels "a bit cleaner" for @NikhilSDate there. In it, `sp.plot(f, g, label=('f','g'), show=False)` will simply extract the two labels from the argument.

What do you think of this?

Who writes the code, decides, of course :-) I will be happily using the API implemented here, when it becomes available.

File: sympy/plotting/plot.py (hunk 1 above)
Comment (smichr) [reply sympy/plotting/plot.py]: We are now at 1.10

**Comment by Davide-sd** [comment]
this PR can be closed. A new plotting module is coming. See https://github.com/sympy/sympy/issues/23036 .

**Pull request closed without merging**